users/engler/fairseq/training.py

```diff
@@ -0,0 +1,128 @@
+        :param str yaml_prefix: Prefix which should be written to the beginning of the config, for example "# @package _group_"
+        """
+        assert isinstance(config_dict, dict)
+        self.config_dict = config_dict
+        self.yaml_prefix = yaml_prefix
+
+    def write(self, path):
+        config_yaml = yaml.dump(self.config_dict)
+        # "# @package _group_" was written at the beginning in the example .yaml from fairseq:
+        if self.yaml_prefix != "":
+            config_yaml = self.yaml_prefix + "\n" + config_yaml
+        with open(path, "w") as file:
+            file.write(config_yaml)
+
+
+class FairseqHydraTrainingJob(Job):
+    """
+    Train a Fairseq model using fairseq-hydra-train
+    """
+
+    def __init__(
+        self,
+        fairseq_hydra_config,
+        *,  # args below are keyword only
+        command_line_args=None,
+        max_epoch=1,
+        save_interval=1,
+        keep_epochs=None,
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=2,
+        gpu_rqmt=1,
+        fairseq_python_exe=None,
+        fairseq_hydra_exe=None,
+    ):
+        """
+        :param FairseqHydraConfig fairseq_hydra_config:
+        :param list command_line_args: Additional command line arguments (starting with "--*") to configure the Fairseq-hydra task
+        :param int max_epoch: maximum number of epochs to run. Note that this value IS currently HASHED.
+        :param int save_interval: save a checkpoint each n-th epoch
+        :param list[int]|set[int]|None keep_epochs: specify which checkpoints are kept in self.out_models, use None for each save_interval-th epoch
+        :param int|float time_rqmt: Overall time requirements
+        :param int|float mem_rqmt: Memory requirements (per GPU)
+        :param int cpu_rqmt: Required number of CPUs (per GPU)
+        :param int gpu_rqmt: Number of required GPUs
+        :param Path|str python_exe: File path to the executable for running python
+        :param Path|str fairseq_hydra_exe: File path to the Fairseq executable: fairseq-hydra-train (usually in the same folder as python exe '.../bin/')
+
+        """
+        self.fairseq_hydra_config = fairseq_hydra_config
+        self.command_line_args = command_line_args or []
+
+        stored_epochs = list(range(save_interval, max_epoch, save_interval)) + [
+            max_epoch
+        ]
+        if keep_epochs is None:
+            self.keep_epochs = set(stored_epochs)
+        else:
+            self.keep_epochs = set(keep_epochs)
+        self.max_epoch = max_epoch
+        self.save_interval = save_interval
+        self.out_fairseq_hydra_yaml = self.output_path("fairseq_hydra_config.yaml")
+        self.out_checkpoint_dir = self.output_path("checkpoints", directory=True)
+        self.out_models = {
+            k: str(self.out_checkpoint_dir) + "/checkpoint{}.pt".format(k)
```

Comment:
Attention: this will write the absolute path of the model files in this dict. Therefore each subsequent Job using an item of this dict 
a) will have a hash depending on the location of your setup
b) will not know that the model file is linked to this training Job. So the dependency graph will be broken here

I suggest using a checkpoint class similar to https://github.com/rwth-i6/i6_core/blob/d27320747b0d7ebe40b8fec6a96cdacb7a961aa5/returnn/training.py#L38-L61

Reply:
good point! thank you

@JanEnglerRWTH this would be a very useful addition